README.md
![](https://img.shields.io/badge/Coverage-93%25-83A603.svg?prefix=$coverage$)
[![pipeline status](https://github.com/art0rz/foxhole-warapi/actions/workflows/release.yml/badge.svg)](https://github.com/github/docs/actions/workflows/main.yml/badge.svg)
[![npm version](https://badge.fury.io/js/foxhole-warapi.svg)](https://www.npmjs.com/package/foxhole-warapi)
# Foxhole WarAPI utilities and TypeScript types
This library provides several utility functions for working with the Foxhole WarAPI. It also includes TypeScript types.

## Getting started
This module has been published on npm. Install it by running the following command:

```shell
npm install foxhole-warapi
```

## Docs
See [Documentation](docs/README.md)

CHANGELOG.md
# [1.7.0](https://github.com/art0rz/foxhole-warapi/compare/v1.6.0...v1.7.0) (2023-02-27)


### Features

* include date in Event ([52442fe](https://github.com/art0rz/foxhole-warapi/commit/52442fe54dbe52cc9372e0251d41fc2d7c591c74))

# [1.6.0](https://github.com/art0rz/foxhole-warapi/compare/v1.5.8...v1.6.0) (2023-02-27)


### Features

* include iconType/text/x/y in Event type ([f01d4a6](https://github.com/art0rz/foxhole-warapi/commit/f01d4a603259d821f1232cf97f6d562d28af4941))

## [1.5.8](https://github.com/art0rz/foxhole-warapi/compare/v1.5.7...v1.5.8) (2023-02-24)


### Bug Fixes

* **build:** include source maps in package ([e3ebc34](https://github.com/art0rz/foxhole-warapi/commit/e3ebc34e2418f04b36215611434989eb4ab6b7ec))

## [1.5.7](https://github.com/art0rz/foxhole-warapi/compare/v1.5.6...v1.5.7) (2023-02-24)

## [1.5.6](https://github.com/art0rz/foxhole-warapi/compare/v1.5.5...v1.5.6) (2023-02-23)


### Bug Fixes

* **determineEventStatusType:** improve event type matching ([d35321f](https://github.com/art0rz/foxhole-warapi/commit/d35321f5f25ac98be9ab90c465b28cd573d679fa))

## [1.5.5](https://github.com/art0rz/foxhole-warapi/compare/v1.5.4...v1.5.5) (2023-02-23)


### Bug Fixes

* **determineEventStatusType:** improve event type matching ([faccaba](https://github.com/art0rz/foxhole-warapi/commit/faccabad079f767e7e0fee366a4b93b91d491951))

## [1.5.4](https://github.com/art0rz/foxhole-warapi/compare/v1.5.3...v1.5.4) (2023-02-23)


### Bug Fixes

* **determineEventStatusType:** add `Upgraded` status ([a29aebc](https://github.com/art0rz/foxhole-warapi/commit/a29aebcc846b8740f5213b2c9912c8ba34836a2f))

## [1.5.3](https://github.com/art0rz/foxhole-warapi/compare/v1.5.2...v1.5.3) (2023-02-23)


### Bug Fixes

* **determineEventStatusType:** add `Upgrading` status ([f2ea5d4](https://github.com/art0rz/foxhole-warapi/commit/f2ea5d4c5952b04b5717264c6cb9460f2b5e4450))

## [1.5.2](https://github.com/art0rz/foxhole-warapi/compare/v1.5.1...v1.5.2) (2023-02-23)


### Bug Fixes

* **tests:** turn on collectCoverage ([010af4a](https://github.com/art0rz/foxhole-warapi/commit/010af4a03f3e6771dfc8909670ba011f07329f5f))

## [1.5.1](https://github.com/art0rz/foxhole-warapi/compare/v1.5.0...v1.5.1) (2023-02-23)


### Bug Fixes

* **tests:** turn on collectCoverage ([04a2bac](https://github.com/art0rz/foxhole-warapi/commit/04a2bac1ef5f1f870938be7bb1f95892fe9acc57))

# [1.5.0](https://github.com/art0rz/foxhole-warapi/compare/v1.4.0...v1.5.0) (2023-02-23)


### Features

* **pipeline:** add status badges ([0c8a76c](https://github.com/art0rz/foxhole-warapi/commit/0c8a76cff19649e55c0032cdff979f37087ee24d))

# [1.4.0](https://github.com/art0rz/foxhole-warapi/compare/v1.3.1...v1.4.0) (2023-02-23)


### Features

* **pipeline:** add status badges ([ec0f94d](https://github.com/art0rz/foxhole-warapi/commit/ec0f94df866ce927910d02efbc6bb523371162c5))

## [1.3.1](https://github.com/art0rz/foxhole-warapi/compare/v1.3.0...v1.3.1) (2023-02-23)


### Bug Fixes

* **pipeline:** make semantic-release also commit new docs ([036ec6a](https://github.com/art0rz/foxhole-warapi/commit/036ec6a6d9967d31ef89f3d8c28e131344b4e3ec))

# [1.3.0](https://github.com/art0rz/foxhole-warapi/compare/v1.2.2...v1.3.0) (2023-02-23)


### Features

* add functions for diffing old and new data and getting event type ([5961b13](https://github.com/art0rz/foxhole-warapi/commit/5961b13f5384d6a67345aa161a4f01a867a11ff5))

## [1.2.2](https://github.com/art0rz/foxhole-warapi/compare/v1.2.1...v1.2.2) (2023-02-22)

## [1.2.1](https://github.com/art0rz/foxhole-warapi/compare/v1.2.0...v1.2.1) (2023-02-22)


### Bug Fixes

* **build:** change build directory ([a43971e](https://github.com/art0rz/foxhole-warapi/commit/a43971e7f0ae2276816a8613187f8b1ea466f20c))
* **build:** change build directory ([fa800ea](https://github.com/art0rz/foxhole-warapi/commit/fa800ea80eaf0beb9f234dacda2fc36bf0a347ae))

# [1.2.0](https://github.com/art0rz/foxhole-warapi/compare/v1.1.0...v1.2.0) (2023-02-22)


### Features

* **types:** add additional types ([85ada9f](https://github.com/art0rz/foxhole-warapi/commit/85ada9ffce1e7e490745485f8ac9424147f4fca1))

# [1.1.0](https://github.com/art0rz/foxhole-warapi/compare/v1.0.2...v1.1.0) (2023-02-22)


### Features

* add commitlint ([1303114](https://github.com/art0rz/foxhole-warapi/commit/130311493809c4cdcd61c66585a5e01588070a53))
* add lint-staged with eslint setup ([bc9194f](https://github.com/art0rz/foxhole-warapi/commit/bc9194fdec76d9a1f79af5af0d4341013df50a26))

## [1.0.2](https://github.com/art0rz/foxhole-warapi/compare/v1.0.1...v1.0.2) (2023-02-22)


### Bug Fixes

* add @semantic-release/npm ([02a6819](https://github.com/art0rz/foxhole-warapi/commit/02a6819a9738fc77e3be3995babde9217d755f30))

## [1.0.1](https://github.com/art0rz/foxhole-warapi/compare/v1.0.0...v1.0.1) (2023-02-22)


### Bug Fixes

* remove docs pipeline as it was interfering with semantic-release ([b8b3ce9](https://github.com/art0rz/foxhole-warapi/commit/b8b3ce9dd0edb7d50cb42b1177899d7b04006c5b))

# 1.0.0 (2023-02-22)


### Bug Fixes

* add semantic-release dependencies ([329f4b6](https://github.com/art0rz/foxhole-warapi/commit/329f4b63acaa2fb6669d4e6f6db227093612be67))
* improve rollup build config, add terser ([3b61061](https://github.com/art0rz/foxhole-warapi/commit/3b610611aed14c4cadee935b688f82134c33d85d))


### Features

* add automatic releases with semantic-release ([4a63186](https://github.com/art0rz/foxhole-warapi/commit/4a631862ea46c829b8d9b834b0e29a85a0f80e98))
* add automatic releases with semantic-release ([d057239](https://github.com/art0rz/foxhole-warapi/commit/d0572391b99d9cefbfba51fc5c54091988d79072))


src/math.test.ts
import { distanceBetweenPoints } from './math';

describe('distanceBetweenPoints', () => {
	it('should calculate the distance between two points correctly', () => {
		expect(
			distanceBetweenPoints(
				{
					x: 0,
					y: 0,
				},
				{
					x: 1,
					y: 0,
				}
			)
		).toEqual(1);
		expect(
			distanceBetweenPoints(
				{
					x: 0,
					y: 0,
				},
				{
					x: 0,
					y: 1,
				}
			)
		).toEqual(1);
	});
});


src/war-api.test.ts
import {
	combineDynamicWithStaticData,
	determineEventStatusType,
	diffCombinedMapData,
	findNearestTextItem,
	getMapFlags,
} from './war-api';
import { EventType, MapFlags } from './types';

function determineEventStatusTypeExcludingDate(
	...args: Parameters<typeof determineEventStatusType>
): Omit<ReturnType<typeof determineEventStatusType>, 'date'> {
	return {
		...determineEventStatusType(...args),
		date: undefined,
	};
}

describe('findNearestTextItem', () => {
	it('should find the nearest item', () => {
		expect(
			findNearestTextItem(
				{
					x: 1,
					y: 1,
					teamId: 'NONE',
					flags: 0,
					iconType: 1,
				},
				[
					{
						x: 1.1,
						y: 1.1,
						text: 'Text 1',
						mapMarkerType: 'Major',
					},
					{
						x: 2.1,
						y: 2.1,
						text: 'Text 2',
						mapMarkerType: 'Major',
					},
				]
			).text
		).toEqual('Text 1');
		expect(
			findNearestTextItem(
				{
					x: 2,
					y: 2,
					teamId: 'NONE',
					flags: 0,
					iconType: 1,
				},
				[
					{
						x: 1.1,
						y: 1.1,
						text: 'Text 1',
						mapMarkerType: 'Major',
					},
					{
						x: 2.1,
						y: 2.1,
						text: 'Text 2',
						mapMarkerType: 'Major',
					},
				]
			).text
		).toEqual('Text 2');
	});
});

describe('combineDynamicWithStaticData', () => {
	it('should combine mapItems with mapTextItems correctly', () => {
		const res = combineDynamicWithStaticData(
			[
				{
					x: 1,
					y: 1,
					teamId: 'NONE',
					flags: 0,
					iconType: 1,
				},
				{
					x: 2,
					y: 2,
					teamId: 'NONE',
					flags: 0,
					iconType: 1,
				},
			],
			[
				{
					x: 1.1,
					y: 1.1,
					text: 'Text 1',
					mapMarkerType: 'Major',
				},
				{
					x: 2.1,
					y: 2.1,
					text: 'Text 2',
					mapMarkerType: 'Major',
				},
			]
		);
		expect(res[0].mapTextItem.text).toEqual('Text 1');
		expect(res[1].mapTextItem.text).toEqual('Text 2');
	});
});

describe('getMapFlags', () => {
	it('should get the right map flags', () => {
		expect(getMapFlags(41)).toEqual([MapFlags.IsVictoryBase, MapFlags.IsTownClaimed]);
		expect(getMapFlags(55)).toEqual([
			MapFlags.IsVictoryBase,
			MapFlags.IsHomeBase,
			MapFlags.IsBuildSite,
			MapFlags.IsScorched,
			MapFlags.IsTownClaimed,
		]);
		expect(getMapFlags(52)).toEqual([MapFlags.IsBuildSite, MapFlags.IsScorched, MapFlags.IsTownClaimed]);
	});
});

describe('determineEventStatusType', () => {
	it('should return the correct event type and teamId', () => {
		expect(
			determineEventStatusTypeExcludingDate({
				old: {
					mapItem: {
						teamId: 'COLONIALS',
						iconType: 56,
						x: 0.8525794,
						y: 0.5295663,
						flags: 4,
					},
					mapTextItem: {
						text: 'Tuatha Watchpost',
						x: 0.853922,
						y: 0.5507405,
						mapMarkerType: 'Major',
					},
				},
				new: {
					mapItem: {
						teamId: 'COLONIALS',
						iconType: 57,
						x: 0.8525794,
						y: 0.5295663,
						flags: 8,
					},
					mapTextItem: {
						text: 'Tuatha Watchpost',
						x: 0.853922,
						y: 0.5507405,
						mapMarkerType: 'Major',
					},
				},
			})
		).toEqual({
			event: EventType.Upgraded,
			byTeam: 'COLONIALS',
			iconType: 56,
			text: 'Tuatha Watchpost',
			x: 0.8525794,
			y: 0.5295663,
		});
		expect(
			determineEventStatusTypeExcludingDate({
				old: {
					mapItem: {
						teamId: 'COLONIALS',
						iconType: 45,
						x: 0.6523639,
						y: 0.7225221,
						flags: 8,
					},
					mapTextItem: {
						text: 'Spirit Watch',
						x: 0.6523898,
						y: 0.7255189,
						mapMarkerType: 'Major',
					},
				},
				new: {
					mapItem: {
						teamId: 'NONE',
						iconType: 45,
						x: 0.6523639,
						y: 0.7225221,
						flags: 0,
					},
					mapTextItem: {
						text: 'Spirit Watch',
						x: 0.6523898,
						y: 0.7255189,
						mapMarkerType: 'Major',
					},
				},
			})
		).toEqual({
			event: EventType.Lost,
			byTeam: 'COLONIALS',
			iconType: 45,
			text: 'Spirit Watch',
			x: 0.6523639,
			y: 0.7225221,
		});

		expect(
			determineEventStatusTypeExcludingDate({
				old: {
					mapItem: {
						teamId: 'COLONIALS',
						iconType: 28,
						x: 0.841185,
						y: 0.52723134,
						flags: 0,
					},
					mapTextItem: {
						text: 'Mousetrap',
						x: 0.8401006,
						y: 0.53316337,
						mapMarkerType: 'Major',
					},
				},
				new: {
					mapItem: {
						teamId: 'NONE',
						iconType: 28,
						x: 0.841185,
						y: 0.52723134,
						flags: 0,
					},
					mapTextItem: {
						text: 'Mousetrap',
						x: 0.8401006,
						y: 0.53316337,
						mapMarkerType: 'Major',
					},
				},
			})
		).toEqual({
			event: EventType.Lost,
			byTeam: 'COLONIALS',
			iconType: 28,
			text: 'Mousetrap',
			x: 0.841185,
			y: 0.52723134,
		});

		expect(
			determineEventStatusTypeExcludingDate({
				old: {
					mapItem: {
						teamId: 'COLONIALS',
						iconType: 56,
						x: 0.8392175,
						y: 0.54724175,
						flags: 8,
					},
					mapTextItem: {
						text: 'Mousetrap',
						x: 0.8401006,
						y: 0.53316337,
						mapMarkerType: 'Major',
					},
				},
				new: {
					mapItem: {
						teamId: 'NONE',
						iconType: 56,
						x: 0.8392175,
						y: 0.54724175,
						flags: 0,
					},
					mapTextItem: {
						text: 'Mousetrap',
						x: 0.8401006,
						y: 0.53316337,
						mapMarkerType: 'Major',
					},
				},
			})
		).toEqual({
			event: EventType.Lost,
			byTeam: 'COLONIALS',
			iconType: 56,
			text: 'Mousetrap',
			x: 0.8392175,
			y: 0.54724175,
		});

		expect(
			determineEventStatusTypeExcludingDate({
				old: {
					mapItem: {
						teamId: 'NONE',
						iconType: 40,
						x: 0.11745039,
						y: 0.473913,
						flags: 0,
					},
					mapTextItem: {
						text: 'Hungry Wolf',
						x: 0.17645623,
						y: 0.46003953,
						mapMarkerType: 'Major',
					},
				},
				new: {
					mapItem: {
						teamId: 'WARDENS',
						iconType: 40,
						x: 0.11745039,
						y: 0.473913,
						flags: 4,
					},
					mapTextItem: {
						text: 'Hungry Wolf',
						x: 0.17645623,
						y: 0.46003953,
						mapMarkerType: 'Major',
					},
				},
			})
		).toEqual({
			event: EventType.UnderConstruction,
			byTeam: 'WARDENS',
			iconType: 40,
			text: 'Hungry Wolf',
			x: 0.11745039,
			y: 0.473913,
		});

		expect(
			determineEventStatusTypeExcludingDate({
				old: {
					mapItem: {
						teamId: 'WARDENS',
						iconType: 40,
						x: 0.11745039,
						y: 0.473913,
						flags: 4,
					},
					mapTextItem: {
						text: 'Hungry Wolf',
						x: 0.17645623,
						y: 0.46003953,
						mapMarkerType: 'Major',
					},
				},
				new: {
					mapItem: {
						teamId: 'NONE',
						iconType: 40,
						x: 0.11745039,
						y: 0.473913,
						flags: 0,
					},
					mapTextItem: {
						text: 'Hungry Wolf',
						x: 0.17645623,
						y: 0.46003953,
						mapMarkerType: 'Major',
					},
				},
			})
		).toEqual({
			event: EventType.ConstructionCancelled,
			byTeam: 'WARDENS',
			iconType: 40,
			text: 'Hungry Wolf',
			x: 0.11745039,
			y: 0.473913,
		});
	});
});

describe('diffCombinedMapData', () => {
	it('should diff combine map data correctly', () => {
		expect(
			diffCombinedMapData(
				[
					{
						mapItem: {
							teamId: 'WARDENS',
							iconType: 47,
							x: 0.38665664,
							y: 0.30743244,
							flags: 8,
						},
						mapTextItem: {
							text: 'Shattered Advance',
							x: 0.38944197,
							y: 0.30851248,
							mapMarkerType: 'Major',
						},
					},
				],
				[
					{
						mapItem: {
							teamId: 'NONE',
							iconType: 47,
							x: 0.38665664,
							y: 0.30743244,
							flags: 0,
						},
						mapTextItem: {
							text: 'Shattered Advance',
							x: 0.38944197,
							y: 0.30851248,
							mapMarkerType: 'Major',
						},
					},
				]
			)
		).toEqual([
			{
				old: {
					mapItem: {
						teamId: 'WARDENS',
						iconType: 47,
						x: 0.38665664,
						y: 0.30743244,
						flags: 8,
					},
					mapTextItem: {
						text: 'Shattered Advance',
						x: 0.38944197,
						y: 0.30851248,
						mapMarkerType: 'Major',
					},
				},
				new: {
					mapItem: {
						teamId: 'NONE',
						iconType: 47,
						x: 0.38665664,
						y: 0.30743244,
						flags: 0,
					},
					mapTextItem: {
						text: 'Shattered Advance',
						x: 0.38944197,
						y: 0.30851248,
						mapMarkerType: 'Major',
					},
				},
			},
		]);
	});
});


src/war-api.ts
import { CombinedMapData, DiffedMapData, Event, EventType, MapFlags, MapItem, MapTextItem } from './types';
import { distanceBetweenPoints } from './math';

const mapFlagValues: Record<MapFlags, number> = {
	[MapFlags.IsVictoryBase]: 0x01,
	[MapFlags.IsHomeBase]: 0x02,
	[MapFlags.IsBuildSite]: 0x04,
	[MapFlags.IsScorched]: 0x10,
	[MapFlags.IsTownClaimed]: 0x20,
};

/**
 * Given a MapItem and a list of MapTextItems, finds the nearest MapTextItem. This is useful for determining which town a MapItem belongs to.
 */
export function findNearestTextItem(mapItem: MapItem, mapTextItems: Array<MapTextItem>): MapTextItem {
	const distances = mapTextItems.map(item => distanceBetweenPoints(mapItem, item));
	const closest = [...distances].sort()[0];
	return mapTextItems[distances.indexOf(closest)];
}

/**
 * Matches each MapItem with the closest MapTextItem.
 */
export function combineDynamicWithStaticData(
	mapItems: Array<MapItem>,
	mapTextItems: Array<MapTextItem>
): Array<CombinedMapData> {
	return mapItems
		.map<any>(mapItem => {
			const mapTextItem = findNearestTextItem(mapItem, mapTextItems);
			return {
				mapItem,
				mapTextItem,
			};
		})
		.filter(item => item);
}

/**
 * Returns true if the flags number contains the given MapFlag.
 */
export function hasMapFlag(flags: number, flag: MapFlags): boolean {
	// eslint-disable-next-line no-bitwise
	return (flags & mapFlagValues[flag]) !== 0;
}

/**
 * Returns a list of each MapFlag that the flags contain.
 */
export function getMapFlags(flags: number): Array<MapFlags> {
	// eslint-disable-next-line no-bitwise
	return Object.values(MapFlags).filter(flag => hasMapFlag(flags, flag));
}

/**
 * Returns a list of CombinedMapData where the map flags or team id from the two arguments is different. Assumes that the order of the two lists is the same.
 * @param data1
 * @param data2
 */
export function diffCombinedMapData(
	data1: Array<CombinedMapData>,
	data2: Array<CombinedMapData>
): Array<DiffedMapData> {
	const different = data2.filter((item, index) => {
		return item.mapItem.flags !== data1[index].mapItem.flags || item.mapItem.teamId !== data1[index].mapItem.teamId;
	});

	return different.map(item => ({
		old: data1[data2.indexOf(item)],
		new: item,
	}));
}

export function determineEventStatusType(data: DiffedMapData): Event | undefined {
	const oldFlags = getMapFlags(data.old.mapItem.flags);
	const newFlags = getMapFlags(data.new.mapItem.flags);
	const date = new Date();

	if (
		oldFlags.includes(MapFlags.IsBuildSite) === false &&
		newFlags.includes(MapFlags.IsBuildSite) === true &&
		data.old.mapItem.teamId === 'NONE' &&
		data.new.mapItem.teamId !== 'NONE'
	) {
		return {
			date,
			event: EventType.UnderConstruction,
			byTeam: data.new.mapItem.teamId,
			iconType: data.old.mapItem.iconType,
			x: data.old.mapItem.x,
			y: data.old.mapItem.y,
			text: data.old.mapTextItem.text,
		};
	}

	if (
		oldFlags.includes(MapFlags.IsBuildSite) === true &&
		newFlags.includes(MapFlags.IsBuildSite) === false &&
		data.old.mapItem.teamId !== 'NONE' &&
		data.new.mapItem.teamId !== 'NONE' &&
		data.old.mapItem.iconType !== data.new.mapItem.iconType &&
		data.old.mapItem.teamId === data.new.mapItem.teamId
	) {
		return {
			date,
			event: EventType.Upgraded,
			byTeam: data.new.mapItem.teamId,
			iconType: data.old.mapItem.iconType,
			x: data.old.mapItem.x,
			y: data.old.mapItem.y,
			text: data.old.mapTextItem.text,
		};
	}

	if (
		oldFlags.includes(MapFlags.IsBuildSite) === false &&
		newFlags.includes(MapFlags.IsBuildSite) === false &&
		data.old.mapItem.teamId === 'NONE' &&
		data.new.mapItem.teamId !== 'NONE'
	) {
		return {
			date,
			event: EventType.Taken,
			byTeam: data.new.mapItem.teamId,
			iconType: data.old.mapItem.iconType,
			x: data.old.mapItem.x,
			y: data.old.mapItem.y,
			text: data.old.mapTextItem.text,
		};
	}

	if (
		oldFlags.includes(MapFlags.IsBuildSite) === true &&
		newFlags.includes(MapFlags.IsBuildSite) === false &&
		data.old.mapItem.teamId !== 'NONE' &&
		data.old.mapItem.teamId === data.new.mapItem.teamId
	) {
		return {
			date,
			event: EventType.Taken,
			byTeam: data.new.mapItem.teamId,
			iconType: data.old.mapItem.iconType,
			x: data.old.mapItem.x,
			y: data.old.mapItem.y,
			text: data.old.mapTextItem.text,
		};
	}

	if (
		oldFlags.includes(MapFlags.IsBuildSite) === false &&
		newFlags.includes(MapFlags.IsBuildSite) === false &&
		data.old.mapItem.teamId !== 'NONE' &&
		data.new.mapItem.teamId === 'NONE'
	) {
		return {
			date,
			event: EventType.Lost,
			byTeam: data.old.mapItem.teamId,
			iconType: data.old.mapItem.iconType,
			x: data.old.mapItem.x,
			y: data.old.mapItem.y,
			text: data.old.mapTextItem.text,
		};
	}

	if (
		oldFlags.includes(MapFlags.IsBuildSite) === true &&
		newFlags.includes(MapFlags.IsBuildSite) === false &&
		data.old.mapItem.teamId !== 'NONE' &&
		data.new.mapItem.teamId === 'NONE'
	) {
		return {
			date,
			event: EventType.ConstructionCancelled,
			byTeam: data.old.mapItem.teamId,
			iconType: data.old.mapItem.iconType,
			x: data.old.mapItem.x,
			y: data.old.mapItem.y,
			text: data.old.mapTextItem.text,
		};
	}

	return undefined;
}


src/types.ts
export type TeamId = 'WARDENS' | 'COLONIALS' | 'NONE';

export interface WarInfo {
	warId: string;
	warNumber: number;
	winner: TeamId;
	conquestStartTime: number;
	conquestEndTime: number | null;
	resistanceStartTime: number | null;
	requiredVictoryTowns: number;
}

export interface WarReport {
	totalEnlistments: number;
	colonialCasualties: number;
	wardenCasualties: number;
	dayOfWar: number;
	version: number;
}

export enum MapIconTypes {
	Static_Base_1 = 5, // Removed in v0.46
	Static_Base_2 = 6, // Removed in v0.46
	Static_Base_3 = 7, // Removed in v0.46
	Forward_Base_1 = 8,
	Forward_Base_2 = 9, // Removed in v0.50
	Forward_Base_3 = 10, // Removed in v0.50
	Hospital = 11,
	Vehicle_Factory = 12,
	Armory = 13,
	Supply_Station = 14,
	Workshop = 15,
	Manufacturing_Plant = 16,
	Refinery = 17,
	Shipyard = 18,
	Tech_Center = 19, // (Engineering Center in v0.37)
	Salvage_Field = 20,
	Component_Field = 21,
	Fuel_Field = 22,
	Sulfur_Field = 23,
	World_Map_Tent = 24,
	Travel_Tent = 25,
	Training_Area = 26,
	Special_Base_Keep = 27, // v0.14
	Observation_Tower = 28, // v0.14
	Fort = 29, // v0.14
	Troop_Ship = 30, // v0.14
	Sulfur_Mine = 32, // v0.16
	Storage_Facility = 33, // v0.17
	Factory = 34, // v0.17
	Garrison_Station = 35, // v0.20
	Ammo_Factory = 36, // v0.20
	Rocket_Site = 37, // v0.20
	Salvage_Mine = 38, // v0.22
	Construction_Yard = 39, // v0.26
	Component_Mine = 40, // v0.26
	Oil_Well = 41, // v0.26 // Removed in v0.50
	Relic_Base_1 = 45, // v0.32
	Relic_Base_2 = 46, // v0.32
	Relic_Base_3 = 47, // v0.32
	Mass_Production_Factory = 51, // v0.35
	Seaport = 52, // v0.37
	Coastal_Gun = 53, // v0.37
	Soul_Factory = 54, // v0.39
	Town_Base_1 = 56, // v0.46
	Town_Base_2 = 57, // v0.46
	Town_Base_3 = 58, // v0.46
	Storm_Cannon = 59, // v0.47
	Intel_Center = 60, // v0.47
	Coal_Field = 61, // v0.50
	Oil_Field = 62, // v0.50
}

export enum MapFlags {
	IsVictoryBase = 'IsVictoryBase',
	IsHomeBase = 'IsHomeBase', // Removed in v0.29
	IsBuildSite = 'IsBuildSite',
	IsScorched = 'IsScorched', // v0.22
	IsTownClaimed = 'IsTownClaimed', // v0.26
}

export interface MapItem {
	teamId: TeamId;
	iconType: MapIconTypes;
	x: number;
	y: number;
	flags: number;
}

export interface MapData {
	regionId: number;
	scorchedVictoryTowns: number;
	mapItems: Array<MapItem>;
	mapTextItems: Array<MapTextItem>;
	lastUpdated: number;
	version: number;
}

export type MapMarkerType = 'Major' | 'Minor';

export interface MapTextItem {
	text: string;
	x: number;
	y: number;
	mapMarkerType: MapMarkerType;
}

export interface CombinedMapData {
	mapItem: MapItem;
	mapTextItem: MapTextItem;
}

export interface DiffedMapData {
	old: {
		mapItem: MapItem;
		mapTextItem: MapTextItem;
	};
	new: {
		mapItem: MapItem;
		mapTextItem: MapTextItem;
	};
}

export enum EventType {
	Lost = 'Lost',
	UnderConstruction = 'UnderConstruction',
	ConstructionCancelled = 'ConstructionCancelled',
	Taken = 'Taken',
	Upgraded = 'Upgraded',
}

export interface Event {
	date: Date;
	event: EventType;
	byTeam: TeamId;
	iconType: number;
	text: string;
	x: number;
	y: number;
}
